src/Processors/Transforms/CountingTransform.cpp

```diff
@@ -18,11 +18,17 @@ namespace DB
 
 void CountingTransform::onConsume(Chunk chunk)
 {
+    if (quota)
+    {
+        /// Do not throw if quota exceded right now, because
```

Comment:
But why not to throw early?
So, if a user is going to overdraft the quota, exception will be thrown before overdrafting.

Reply:
The idea is that bytes are not written yet (in comparison with `READ_BYTES` quota, where bytes have been already read at the moment of check) and there is only an intent to write bytes. It will be better to check whether quota will be exceed if write that amount of data, and if it's exceeded throw without incrementing usage, but it's a bit harder to implement, because it's easy to add a race with such logic.

Reply:
Also consider a situation when we have a 1Gb quota for written bytes and user sends an insert with 1.1Gb of data. The quota will be exceeded, but data won't be written to the table and user will be unable to send, e.g 500Mb of data.

But for simplicity we can make this quota more strict as you suggest.

Reply:
Yes, let's make it more strict. When even an intent to write something will count.